src/autorest.bicep/src/schema-generator.ts

```diff
@@ -0,0 +1,729 @@
+import { chain, Dictionary, escapeRegExp, keys, orderBy, uniq } from 'lodash';
+import { getFullyQualifiedType, getNameSchema, getSerializedName, NameSchema, ProviderDefinition, ResourceDefinition, ResourceDescriptor } from "./resources";
+import { isEmpty } from 'lodash';
+import { ScopeType } from "bicep-types";
+import { AnyObjectSchema, AnySchema, ArraySchema, ByteArraySchema, ChoiceSchema, ComplexSchema, ConstantSchema, DateTimeSchema, DictionarySchema, NumberSchema, ObjectSchema, PrimitiveSchema, Property, Schema, SchemaType, SealedChoiceSchema, StringSchema, UuidSchema } from '@autorest/codemodel';
+import { failure, success } from './utils';
+
+interface SchemaData {
+  definitions: Dictionary<JSONSchema4>;
+  rgResources: Dictionary<JSONSchema4>;
+  subResources: Dictionary<JSONSchema4>;
+  mgResources: Dictionary<JSONSchema4>;
+  tenantResources: Dictionary<JSONSchema4>;
+  extensionResources: Dictionary<JSONSchema4>;
+  unknownResources: Dictionary<JSONSchema4>;
+}
+
+export function generateSchema(host: AutorestExtensionHost, definition: ProviderDefinition): JSONSchema4 {
+  const schemaData: SchemaData = {
+    definitions: {},
+    rgResources: {},
+    subResources: {},
+    mgResources: {},
+    tenantResources: {},
+    extensionResources: {},
+    unknownResources: {},
+  }
+
+  function logWarning(message: string) {
+    host.message({ Channel: Channel.Warning, Text: message, });
+  }
+
+  function getResourcePath(definition: ResourceDefinition) {
+    return (definition.putOperation ?? definition.getOperation)?.request.path;
+  }
+
+  function tryGetConstantName(nameSchema: NameSchema) {
+    if (nameSchema.type === 'constant') {
+      return nameSchema.value;
+    }
+
+    if ((nameSchema.schema instanceof ChoiceSchema ||  nameSchema.schema instanceof SealedChoiceSchema) &&
+      nameSchema.schema.choices.length === 1 && typeof nameSchema.schema.choices[0].value === 'string')
+    {
+      return nameSchema.schema.choices[0].value;
+    }
+
+    return;
+  }
+
+  function getResourceNameSchema(descriptor: ResourceDescriptor, nameSchema: NameSchema, isChildDefinition: boolean): JSONSchema4 {
+    const constName = tryGetConstantName(nameSchema);
+
+    if (constName) {
+      if (descriptor.typeSegments.length < 2 || isChildDefinition) {
+        return addExpressionOneOf({
+          type: 'string',
+          enum: [constName],
+        });
+      } else {
+        return addExpressionOneOf({
+          type: 'string',
+          pattern: `^.*/${escapeRegExp(constName)}$`,
+        });
+      }
+    }
+
+    return (nameSchema.type == 'parameterized' ? parseType(nameSchema.schema, true) : undefined) ?? { type: 'string' };
+  }
+
+  function createObject(properties: Dictionary<JSONSchema4>, additionalProperties?: JSONSchema4): JSONSchema4 {
+    return {
+      type: 'object',
+      properties,
+      additionalProperties,
+    };
+  }
+
+  function throwIfNull<TSchema extends Schema>(putSchema: TSchema | undefined) {
```

Comment:
is `TSchema | undefined` different from `TSchema?`?

Reply:
In terms of how the function is processed, I don't think so. It does make the argument optional, which is a bit odd - e.g. you could call `throwIfNull()`